sklearn/decomposition/nmf.py

```diff
@@ -254,22 +344,287 @@ def _nls_subproblem(V, W, H, tol, max_iter, sigma=0.01, beta=0.1):
+        Set to False, only W will be estimated.
+
+    verbose : integer, default: 0
+        The verbosity level.
+
+    shuffle : boolean, default: False
+        If True, the samples will be taken in shuffled order during
+        coordinate descent.
+
+    random_state : integer seed, RandomState instance, or None (default)
+        Random number generator seed control.
+
+    Returns
+    -------
+    W : array-like, shape (n_samples, n_components)
+        Solution to the non-negative least squares problem.
+
+    H : array-like, shape (n_components, n_features)
+        Solution to the non-negative least squares problem.
+
+    n_iter : int
+        The number of iterations done by the algorithm.
 
-    init :  'nndsvd' |  'nndsvda' | 'nndsvdar' | 'random'
+    References
+    ----------
+    Hsieh, Cho-Jui, and Inderjit S. Dhillon. "Fast coordinate descent methods
+    with variable selection for non-negative matrix factorization."
+    Proceedings of the 17th ACM SIGKDD. ACM, 2011.
+    """
+    # so W and Ht are both in C order in memory
+    Ht = check_array(H.T, order='C')
+    X = check_array(X, accept_sparse='csr')
+
+    alpha_H = 0.
+    alpha_W = 0.
+    if regularization in ('both', 'components'):
+        alpha_H = alpha
+    if regularization in ('both', 'transformation'):
+        alpha_W = alpha
+
+    rng = check_random_state(random_state)
+
+    for n_iter in range(max_iter):
+            v = 0.
+
+            # Update W
+            v += _update_coordinate_descent(X, W, Ht, alpha_W, l1_ratio,
+                                            shuffle, rng)
+            # Update H
+            if update_H:
+                v += _update_coordinate_descent(X.T, Ht, W, alpha_H, l1_ratio,
+                                                shuffle, rng)
+
+            if n_iter == 0:
+                violation_init = v
+
+            if violation_init == 0:
+                break
+
+            if verbose:
+                print("violation:", v / violation_init)
+
+            if v / violation_init <= tol:
+                if verbose:
+                    print("Converged at iteration", n_iter + 1)
+                break
+
+    return W, Ht.T, n_iter
+
+
+def non_negative_matrix_factorization(X, W=None, H=None, n_components=None,
+                                      init='random', update_H=True,
+                                      solver='proj-grad', tol=1e-4,
+                                      max_iter=200, alpha=0., l1_ratio=0.,
+                                      regularization=None, random_state=None,
+                                      verbose=0, shuffle=False,
+                                      nls_max_iter=2000, sparseness=None,
+                                      beta=1, eta=0.1):
+    """Compute Non-negative Matrix Factorization (NMF)
+
+    Find (W, H) that minimize || X - W.H ||_2
+    with an alternative least square minimization of W and H.
```

Comment:
same observation about `least square`. Also if you decide to leave it in I think it should be `least squares`.

As in my top-level comment I think the full loss function should be here. (No need to bother with sparseness, beta and eta, they're deprecated, but we could work out what loss function they correspond to.)